hudi-common/src/main/java/org/apache/hudi/common/table/read/IncrementalQueryAnalyzer.java

```diff
@@ -0,0 +1,428 @@
+      boolean skipCompaction,
+      boolean skipClustering,
+      int limit) {
+    this.metaClient = metaClient;
+    this.startTime = Option.ofNullable(startTime);
+    this.endTime = Option.ofNullable(endTime);
+    this.rangeType = rangeType;
+    this.skipCompaction = skipCompaction;
+    this.skipClustering = skipClustering;
+    this.limit = limit;
+  }
+
+  /**
+   * Returns a builder.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * Analyzes the incremental query context with given completion time range.
+   *
+   * @return An incremental query context including the instant time range info.
+   */
+  public QueryContext analyze() {
+    try (CompletionTimeQueryView completionTimeQueryView = new CompletionTimeQueryView(this.metaClient)) {
+      if (completionTimeQueryView.isEmptyTable()) {
+        // no dataset committed in the table
+        return QueryContext.EMPTY;
+      }
+      HoodieTimeline readTimeline = getReadTimeline(this.metaClient);
+      List<String> instantTimeList = completionTimeQueryView.getStartTimes(readTimeline, startTime, endTime, rangeType);
+      if (instantTimeList.isEmpty()) {
+        // no instants completed within the give time range, returns early.
+        return QueryContext.EMPTY;
+      }
+      // get hoodie instants
+      Pair<List<String>, List<String>> splitInstantTime = splitInstantByActiveness(instantTimeList, completionTimeQueryView);
+      Set<String> instantTimeSet = new HashSet<>(instantTimeList);
+      List<String> archivedInstantTime = splitInstantTime.getKey();
+      List<String> activeInstantTime = splitInstantTime.getValue();
+      List<HoodieInstant> archivedInstants = new ArrayList<>();
+      List<HoodieInstant> activeInstants = new ArrayList<>();
+      HoodieTimeline archivedReadTimeline = null;
+      if (!activeInstantTime.isEmpty()) {
+        activeInstants = readTimeline.getInstantsAsStream().filter(instant -> instantTimeSet.contains(instant.getTimestamp())).collect(Collectors.toList());
+        if (limit > 0 && limit < activeInstants.size()) {
+          // streaming read speed limit, limits the maximum number of commits allowed to read for each run
+          activeInstants = activeInstants.subList(0, limit);
+        }
+      }
+      if (!archivedInstantTime.isEmpty()) {
+        archivedReadTimeline = getArchivedReadTimeline(metaClient, archivedInstantTime.get(0));
+        archivedInstants = archivedReadTimeline.getInstantsAsStream().filter(instant -> instantTimeSet.contains(instant.getTimestamp())).collect(Collectors.toList());
+      }
+      List<String> instants = Stream.concat(archivedInstants.stream(), activeInstants.stream()).map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+      if (instants.isEmpty()) {
+        // no instants completed within the give time range, returns early.
+        return QueryContext.EMPTY;
+      }
+      if (startTime.isEmpty() && endTime.isPresent()) {
```

Comment:
Where do we handle both startTime and endTime being empty?

Reply:
L187 should handle `endtTime` being empty?